docs/_howto/link-seams-commonjs.md

```diff
@@ -1,17 +1,80 @@
 ---
 layout: page
-title: How to use Link Seams with CommonJS
+title: Dependency mocking with Sinon
 ---
+This page describes how to isolate your system under test, by stubbing out dependencies.
 
-This page describes how to isolate your system under test, by stubbing out dependencies with [link seams][seams].
+While in other languages you might use [link seams][seams], [Dependency Injection(DI)][di] or
+[Inversion or Control(IoT)][IoT], which are not just _patterns_ but also requires some pattern
+implementation(a framework) - CommonJS(aka nodejs) module system provides a way better to replace
```

Comment:
You don't need a framework to do DI. I do it all the time by having optional constructor parameters. If using links seams really is "better" is debatable, as it introduces a new dependency, apis, complexity and mental overhead, but at least it doesn't require you to change the [SUT](https://en.wikipedia.org/wiki/System_under_test).

In general, you don't need any custom framework to use IoC (of which DI is one pattern). Constructor injection or service locators are simple to implement in any language. Less mystery is one of our tenants, in any case, so we wouldn't want to give people the impression you need special tools, unless you are doing "special" stuff like hooking into the module loading system (as we are doing here).

Reply:
- DI. Accepting your dependencies via arguments, ie using some sort of a factory pattern(and able to produce multiple objects). That's more about runtime.
- dependency mocking. Statically declaring your dependencies, without the ability to factor different versions. That's only about testing.

I would say - DI needs some framework, as long as you have to structure and design your code in a specific way, while dependency mocking operates on a bit different level and in a different time(tests only).

To be honest - I am quite happy that we have such ability in JS as a dependency mocking, so I don't have to implement 3 tier architecture to be able to mock `fs`. That makes code simpler and tests simpler.

There are more Java developers around me than JS, with all that _Enterprise stuff_ from the fairy tales, and I like to show them how simple the code could be. (Sometime they do the same)

Reply:
I agree with the opinion that dependency injection does not need a framework in JavaScript. I often use Sinon's spies, stubs and fakes in combination with pure functions. This makes my code easy enough to reason about, that I can get my colleagues to 👍 my pull requests.